src/hotspot/share/cds/filemap.cpp

```diff
@@ -188,6 +195,22 @@ FileMapInfo::~FileMapInfo() {
     assert(_dynamic_archive_info == this, "must be singleton"); // not thread safe
     _dynamic_archive_info = NULL;
   }
+  if (_file_open) {
+    os::close(_fd);
+  }
+}
+
+// Do preliminary validation on archive. More checks are in initialization.
+bool FileMapInfo::validate_archive() const {
+  if (!os::file_exists(_full_path)) {
+    return false;
+  }
```

Comment:
Is the above check necessary? If the file doesn't exist, will `check_archive()` below return false?

Reply:
It is needed here --- when open a file of name is NULL, os::open first does is checking:
if (strlen(path) > MAX_PATH - 1) {
    errno = ENAMETOOLONG;
    return -1;
  }
It will crash on NULL. Removing it will cause many tests fail.  os:file_exists will first check if it is NULL. 
I will add comment for it.

Reply:
I think we shouldn't be checking a FileMapInfo when its _file_path is NULL. I added the following assert:

```
bool FileMapInfo::validate_archive() const {
+ assert(_full_path != NULL, "huh");
  if (!os::file_exists(_full_path)) {
    return false;
  }
```

and it crashes here (with dynamicArchive/HelloDynamic.java): 

```
 9 FileMapInfo::validate_archive 
10 FileMapInfo::FileMapInfo 
11 DynamicArchiveBuilder::init_header 
12 DynamicArchiveBuilder::doit 
13 VM_PopulateDynamicDumpSharedSpace::doit 
14 VM_Operation::evaluate 
15 VMThread::evaluate_operation 
16 VMThread::inner_execute 
17 VMThread::loop 
...
```

At this point, we are writing the archive, so we shouldn't check the validity of the (output) archive, because we are going to overwrite it. And why is it NULL?
 
By the way, you added FileMapInfo::validate_archive() for doing this inside FileMapInfo::FileMapInfo():

```
  bool valid = validate_archive();
  if (!valid && !_is_static && AutoCreateSharedArchive) {
    // regenerate shared archive at exit
    DynamicDumpSharedSpaces = true;
    ArchiveClassesAtExit = _full_path;
  }
```

But the above block of code doesn't seem necessary, as it's already checked here inside FileMapInfo::initialize():

```
      FileMapInfo::fail_continue("Initialize dynamic archive failed.");
      if (AutoCreateSharedArchive) {
        DynamicDumpSharedSpaces = true;
        ArchiveClassesAtExit = Arguments::GetSharedDynamicArchivePath();
      }
      return false;
```